### PR TITLE
Refactor enginx script to call UI code

### LINKED ISSUE #32663: Update ENGINX script tests to run same code as UI

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Rewrite EnginX script to call GUI code so 

1. Have system test coverage for GUI code
2. Can be used in autoreduction (and will give same answer as GUI)

A couple of changes to an underlying algorithm PDCalibration:

- Fixed bug in PDCalibration such that it now correctly minimises chi-squared (cause of #33046) 
- Added ability to specify fit range for each peak rather than a global window in d-spacing.

**To test:**

(1) Run the systems test script on a non-windows machine to check it passes OK
`systemtest.bat -C debug -R EnginXScriptTest -j 2`

Fixes #32663 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DannyHindson — COMMENTED** [review]
Just had an initial look at this. It's still failing a couple of system tests - maybe the usual troubles getting the RHEL7 ref files set up?

I've got a couple of general questions:
- there are some CMake options that allow Mantid to be built without the qt element eg ENABLE_WORKBENCH and a newer option called MANTID_QT_LIB (which can both be given values OFF). I'm wondering if the EnginX scripting interface into the EnginX code will work with these options set to OFF if it relies on calling the UI model code?
- You've deleted quite a lot of code from the EnggUtils.py unit. Is this because it already exists elsewhere from earlier PRs?

**Comment by RichardWaiteSTFC** [comment]
> Just had an initial look at this. It's still failing a couple of system tests - maybe the usual troubles getting the RHEL7 ref files set up?
> 

I think actually it is due to your diffraction focusing step-size PR #32767 (it was passing the RHEL build before that and then I had a conflict I had to fix that merged in your change). I'm updating the test files now with the build mismatch (but will obviously test on windows as well).

> I've got a couple of general questions:
> 
> * there are some CMake options that allow Mantid to be built without the qt element eg ENABLE_WORKBENCH and a newer option called MANTID_QT_LIB (which can both be given values OFF). I'm wondering if the EnginX scripting interface into the EnginX code will work with these options set to OFF if it relies on calling the UI model code?

I run the tests OK with the systemstest.bat - does that mean it is working? I have MANTID_QT_LIB set to BUILD and ENABLE_WORKBENCH set to True in my cmake.

> * You've deleted quite a lot of code from the EnggUtils.py unit. Is this because it already exists elsewhere from earlier PRs?

Yes I removed some code in PR #32516 but couldn't remove all of it as it was still called in the systems test. A lot of the functionality is now in the class ` CalibrationInfo` (which saves having to pass loads of arguments around etc.)

**Comment by DannyHindson** [comment]
Apologies for the ref files moving. Yes I remember changing these now

Re the cmake options - I think it's all fine with the default options. My question is about whether the EnginX script route into the EnginX code still works OK if you build with some non-default options. So if you set one or both of these to OFF. This might be how the auto-reduction code is built for example. Even if it's not it should probably work still with both of these set to OFF

**Comment by DanielMurphy22** [comment]
I confirm that the new EnginXScriptTest passes locally on MacOS BigSur for me!

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/calibration/model.py [hunk 1]
```diff
@@ -171,8 +171,8 @@ def run_calibration(ceria_ws, calibration, full_instrument_cal_ws):
         cal_table, diag_ws, mask = PDCalibration(InputWorkspace=foc_name, OutputCalibrationTable=cal_table_name,
                                                  DiagnosticWorkspaces=diag_ws_name,
                                                  PeakPositions=EnggUtils.default_ceria_expected_peaks(final=True),
-                                                 TofBinning=[15500, -0.0003, 52000],
-                                                 PeakWindow=0.04,
+                                                 TofBinning=[12000, -0.0003, 52000],
```
Comment (DannyHindson) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:174]: What's the reason for this change to the rebin params?

File: scripts/Engineering/EnggUtils.py [hunk 2]
```diff
@@ -232,179 +93,39 @@ def plot_tof_vs_d_from_calibration(diag_ws, ws_foc, dspacing, calibration):
...
-
-    @param :: final - if true, returns a list better suited to a secondary fitting
+    @param :: final - if true, returns a list better suited to a secondary fitting of focused banks
     @Returns :: a list of peaks in d-spacing as a float list
     """
     _CERIA_EXPECTED_PEAKS = [1.104598643, 1.352851554, 1.631600313, 1.913220892, 2.705702376]
-    _CERIA_EXPECTED_PEAKS_FINAL = [0.855618487, 0.901900955, 0.914694494, 0.956610446, 1.04142562, 1.104598643,
-                                   1.210027059, 1.241461538, 1.352851554, 1.562138267, 1.631600313, 1.913220892,
-                                   2.705702376]
+    _CERIA_EXPECTED_PEAKS_FINAL = [0.781069, 0.855618487, 0.901900955, 0.914694494, 0.956610446, 1.04142562,
```
Comment (DannyHindson) [inline scripts/Engineering/EnggUtils.py:110]: Looks like there's an extra peak in here now? Similar to other question - what's the reason for this?

**Review by DannyHindson — COMMENTED** [review]
I've added a couple of questions but that aside I think this looks fine assuming we pick up the problem about the non-UI build in a separate issue.
Re the questions, just want to record on the PR any other functional changes that have gone in here in addition to the refactor.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/calibration/model.py (hunk 1 above)
Comment (RichardWaiteSTFC) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:174]: The binning occurs in TOF, the original were good for a whole bank but not for some of the texture groups (would chop off the peak at 0.85 Ang in half the spectra). 

image

I then added another peak at lower d-spacing (see below) and so the limit is now considerably lower.

File: scripts/Engineering/EnggUtils.py (hunk 2 above)
Comment (RichardWaiteSTFC) [reply scripts/Engineering/EnggUtils.py:110]: This peak is relatively isolated considering it's at low d-spacing and seemed to be fit well consistently when I was testing so I added it. It should help get TZERO more accurately.

**Review by DannyHindson — APPROVED** [review]
Thanks for the extra detail on the functional changes. This PR will also fix the EnginX system test failure on main so approving now and let's pick up the problem about the EnginXScript unit working for builds with MANTID_QT_LIB=OFF separately

**Comment by RichardWaiteSTFC** [comment]
@mantidproject/gatekeepers I believe this fixes tests that are causing nightly builds to fail on OSX and windows - could it be merged?

**Review by DavidFair — APPROVED** [review]
